shared/util/tf/tf_generator.go

```diff
@@ -0,0 +1,439 @@
+	kubernetesProviderBody.SetAttributeRaw("cluster_ca_certificate", tokens("base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)"))
+	kubernetesProviderBody.SetAttributeRaw("token", tokens("data.aws_eks_cluster_auth.cluster.token"))
+
+	kubeNamespaceBody := rootBody.AppendNewBlock("data", []string{"kubernetes_namespace", "happy-namespace"}).Body()
+	kubeNamespaceBody.AppendNewBlock("metadata", nil).Body().SetAttributeRaw("name", tokens("var.k8s_namespace"))
+
+	awsAliasTokens := tokens("aws.czi-si")
+	appKeyBody := rootBody.AppendNewBlock("data", []string{"aws_ssm_parameter", "dd_app_key"}).Body()
+	appKeyBody.SetAttributeValue("name", cty.StringVal("/shared-infra-prod-datadog/app_key"))
+	appKeyBody.SetAttributeRaw("provider", awsAliasTokens)
+	apiKeyBody := rootBody.AppendNewBlock("data", []string{"aws_ssm_parameter", "dd_api_key"}).Body()
+	apiKeyBody.SetAttributeValue("name", cty.StringVal("/shared-infra-prod-datadog/api_key"))
+	apiKeyBody.SetAttributeRaw("provider", awsAliasTokens)
+
+	datadogProviderBody := rootBody.AppendNewBlock("provider", []string{"datadog"}).Body()
+	datadogProviderBody.SetAttributeRaw("app_key", tokens("data.aws_ssm_parameter.dd_app_key.value"))
+	datadogProviderBody.SetAttributeRaw("api_key", tokens("data.aws_ssm_parameter.dd_api_key.value"))
+
+	_, err = tfFile.Write(hclFile.Bytes())
+
+	return err
+}
+
+func (tf TfGenerator) generateAwsProvider(rootBody *hclwrite.Body, alias, accountIdExpr, roleExpr string) error {
+	awsProviderBody := rootBody.AppendNewBlock("provider", []string{"aws"}).Body()
+	if alias != "" {
+		awsProviderBody.SetAttributeValue("alias", cty.StringVal(alias))
+	}
+	awsProviderBody.SetAttributeValue("region", cty.StringVal(*tf.happyConfig.AwsRegion()))
+
+	assumeRoleBlockBody := awsProviderBody.AppendNewBlock("assume_role", nil).Body()
+	assumeRoleBlockBody.SetAttributeRaw("role_arn", tokens(fmt.Sprintf("\"arn:aws:iam::%s:role/%s\"", accountIdExpr, roleExpr)))
+	awsProviderBody.SetAttributeRaw("allowed_account_ids", tokens(fmt.Sprintf("[\"%s\"]", accountIdExpr)))
+	return nil
+}
+
+func (tf *TfGenerator) GenerateVersions(srcDir string) error {
+	tfFile, err := os.Create(filepath.Join(srcDir, "versions.tf"))
+	if err != nil {
+		return errors.Wrap(err, "Unable to generate HCL code")
+	}
+	defer tfFile.Close()
+	hclFile := hclwrite.NewEmptyFile()
+
+	rootBody := hclFile.Body()
+	terraformBlockBody := rootBody.AppendNewBlock("terraform", nil).Body()
+	terraformBlockBody.SetAttributeValue("required_version", cty.StringVal(requiredTerraformVersion))
+	requiredProvidersBody := terraformBlockBody.AppendNewBlock("required_providers", nil).Body()
+
+	for _, provider := range requiredProviders {
+		p := cty.ObjectVal(map[string]cty.Value{
+			"source":  cty.StringVal(provider.Source),
+			"version": cty.StringVal(provider.Version),
+		})
+		requiredProvidersBody.SetAttributeValue(provider.Name, p)
+
+	}
+
+	_, err = tfFile.Write(hclFile.Bytes())
+
+	return err
```

Comment:
this is a static file, I don't think we need to use an HCL generator for this

Reply:
I need to mull it over.

Reply:
Upon further thought, I'd like to keep it generated. This PR is only the first of the series, this will continue to evolve.